Document post acc grad hooks in backward hooks execution (#107323)
pytorch/pytorch: docs/source/notes/autograd.rst

registered to Node via :meth:`torch.autograd.graph.Node.register_hook`, and
pre-hooks registered to Node via :meth:`torch.autograd.graph.Node.register_prehook`.

Whether a particular hook will be fired
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Hooks registered to a Tensor via :meth:`torch.Tensor.register_hook`
are executed when gradients are being computed for that Tensor. (Note that this does not require
the Tensor's grad_fn to be executed. For example, if the Tensor is passed
as part of the ``inputs`` argument to :func:`torch.autograd.grad`,
the Tensor's grad_fn may not be executed, but the hook register to that Tensor will always be executed.)

Hooks registered to a Tensor via :meth:`torch.Tensor.register_post_accumulate_grad_hook`
are executed after the gradients have been accumulated for that Tensor, meaning the
Tensor's grad field has been set. Whereas hooks registered via :meth:`torch.Tensor.register_hook`
are run as gradients are being computed, hooks registered via :meth:`torch.Tensor.register_post_accumulate_grad_hook`
are only triggered once the Tensor's grad field is updated by autograd at the end of

> the backward pass. Thus, post-accumulate-grad hooks are only fired for leaf Tensors.

Reviewer: Probably want to remove "Thus, post-accumulate-grad hooks are only fired for leaf Tensors." since you could not have registered them to non-leaf tensor anyway.

This section will discuss when different hooks fire or don't fire.
Then it will discuss the order in which they are fired.
The hooks that will be covered are: backward hooks registered to Tensor via

> :meth:`torch.tensor.register_hook`, post accumulate grad hooks registered to

Reviewer 2: torch.Tensor here and below

> :meth:`torch.tensor.register_hook`, post accumulate grad hooks registered to

Reviewer: When referring to "post accumulate grad hooks" I feel like we may need to hyphenate it like "post-accumulate-grad hooks"

as part of the ``inputs`` argument to :func:`torch.autograd.grad`,
the Tensor's grad_fn may not be executed, but the hook register to that Tensor will always be executed.)

Hooks registered to a Tensor via :meth:`torch.tensor.register_post_accumulate_grad_hook`
are executed after the gradients have been accumulated for that Tensor, meaning the
Tensor's grad field has been set. Whereas hooks registered via :meth:`torch.tensor.register_hook`
are run as gradients are being computed, hooks registered via :meth:`torch.tensor.register_post_accumulate_grad_hook`

> wait to trigger on the accumulation of all gradient updates on the Tensor during the backward

Reviewer 2: I don't think this is accurate. There is only ever one call to AccumulateGrad.
The difference I think are:
- register_hook always triggers even when using autograd.grad. Yours ONLY trigger if the .grad field is being updated by autograd.
- the chain of events is register_hook -> update the .grad field -> post acc grad hook

> pass. Similar to node hooks, post accumulate grad hooks are only fired for leaf Tensors for which

Reviewer: "node hooks" -> "hooks registered to Node"

also begin the sentence with "[This means that] similar to node hooks..." since this sentence is implied by the previous.

the AccumulateGrad node has executed. Registering a hook via :meth:`torch.tensor.register_post_accumulate_grad_hook`

> on a non-leaf Tensor will do nothing, even if you call `backward(retain_graph=True)`.

Reviewer: Don't forget to update this, when we change it to error

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The order in which things happen are:

#. hooks registered to Tensor are executed
#. pre-hooks registered to Node are executed (if Node is executed).
#. The ``.grad`` field is updated for Tensors that retain_grad
#. Node is executed (subject to rules above)

> #. If the Node is AccumulateGrad, post accumulate grad hooks registered to Tensor are executed (if Node is executed)

Reviewer: It's tricky to avoid mention of AccumulateGrad Node if we try to explain things this way. I'd be okay with making it a thing in the docs. Though @albanD you mentioned in the other PR that we may want to avoid mention of AccumulateGrad, curious is there's a different preferred way of explaining things like - 
```suggestion
#. for leaf Tensors that have ``.grad`` accumulated, post-accumulate-grad hooks are executed
```

Reviewer 2: I think in the Note that talks about Nodes and how autograd works it is fine.
In the generally user-facing doc I would try to avoid it by just saying "when the .grad field is updated by autograd" as you suggested above

even though the Tensor now has a new grad_fn, hooks registered before it was
modified in-place will continue to be associated with the old grad_fn, e.g. they will
fire when that Tensor's old grad_fn is reached in the graph by the autograd engine.

The one exception is post accumulate grad hooks, which receive the tensor that requires grad
instead of the gradient and expect users to modify the tensor in-place. Ideally, the post
grad accumulate hooks should be last to execute in the backward pass. The only contention
would occur if you also registered a post-hook on the AccumulateGrad node, which is not

> recommended, and in that case, it is up to you to ensure sanity.

Author: I should definitely rewrite this to be like "be cautious and set torch.no_grad for content in the hook if you don't want unforeseen circumstances" and get rid of the "it's last so should be okay". Being last does not make it okay since people run many fwd/bwd passes

Reviewer 2: Note that differentiable inplace on leaf is not allowed. So this comment is moot as you will never be able to register such hook!

Author: So technically...it is "allowed" if  the person enables torch.grad right? Even then, this chunk shouldn't apply to my hooks because I literally give them the Tensor at the time, so any updates from hooks will be reflected.
